resources/net.go

```diff
@@ -0,0 +1,877 @@
+	tempDir string // temporary directory for storing the pipe socket file
+}
+
+// nlChanStruct defines the channel used to send netlink messages and errors
+// to the event processing loop in Watch.
+type nlChanStruct struct {
+	msg []syscall.NetlinkMessage
+	err error
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *NetRes) Default() Res {
+	return &NetRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *NetRes) Validate() error {
+	// validate state
+	if obj.State != ifaceUp && obj.State != ifaceDown && obj.State != "" {
+		return fmt.Errorf("state must be up, down or empty")
+	}
+
+	// validate network address input
+	if (obj.Addrs == nil) != (obj.Gateway == "") {
+		return fmt.Errorf("addrs and gateway must both be set or both be empty")
+	}
+	if obj.Addrs != nil {
+		for _, addr := range obj.Addrs {
+			if _, _, err := net.ParseCIDR(addr); err != nil {
+				return errwrap.Wrapf(err, "error parsing address: %s", addr)
+			}
+		}
+	}
+	if obj.Gateway != "" {
+		if g := net.ParseIP(obj.Gateway); g == nil {
+			return fmt.Errorf("error parsing gateway: %s", obj.Gateway)
+		}
+	}
+
+	// validate the interface name
+	_, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *NetRes) Init() error {
+	var err error
+	// tmp directory for pipe socket
+	// XXX: Replace with obj.VarDir
+	if obj.tempDir, err = ioutil.TempDir("", "pipe"); err != nil {
+		return errwrap.Wrapf(err, "could not get TempDir")
+	}
+
+	// store the network interface in the struct
+	obj.iface = &iface{}
+	if obj.iface.iface, err = net.InterfaceByName(obj.GetName()); err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+	// store the netlink link to use as interface input in netlink functions
+	if obj.iface.link, err = netlink.LinkByName(obj.GetName()); err != nil {
+		return errwrap.Wrapf(err, "error finding link: %s", obj.GetName())
+	}
+
+	// build the path to the networkd configuration file
+	obj.unitFilePath = networkdUnitFileDir + IfacePrefix + obj.GetName() + networkdUnitFileExt
+
+	return obj.BaseRes.Init()
+}
+
+// Watch listens for events from the specified interface via a netlink socket.
+// TODO: currently gets events from ALL interfaces, would be nice to reject
+// events from other interfaces.
+func (obj *NetRes) Watch() error {
+	// waitgroup for netlink receive goroutine
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+
+	// create a netlink socket for receiving network interface events
+	conn, err := newSockSet(rtmGrps, path.Join(obj.tempDir, "pipe.sock"))
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating socket set")
+	}
+	defer conn.shutdown() // close the netlink socket and unblock select
```

Comment:
correct me if i'm wrong, but should this comment be:

"close the netlink socket and unblock conn.receive()"

? (I'm also hoping that's what this does, which if it does, that's great!

Reply:
Yes. It unblocks the select *within* conn.receive(), but in this context, your comment makes more sense.